### PR TITLE
executor, planner: enable duplication check on prefix clustered index

### LINKED ISSUE #18934: clustered index:  broke data consistency when using prefix index primary key
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
use test;
drop table if exists t;
set tidb_enable_clustered_index=1;
create table t(name varchar(255), b int, c int, primary key(name(2)), index idx(b));
insert into t(name, b) values("china", 1), ("sc", 2);
insert into t(name, b) values("cha", 3);
select * from t use index(primary) where name = "china";
```

### 2. What did you expect to see? (Required)
```
mysql> insert into t(name, b) values("china", 1), ("sc", 2);
Query OK, 2 rows affected (0.00 sec)
Records: 2  Duplicates: 0  Warnings: 0

mysql> insert into t(name, b) values("cha", 3);
ERROR 1062 (23000): Duplicate entry 'ch' for key 'PRIMARY'
mysql> select * from t use index(primary) where name = "china";
+-------+------+------+
| name  | b    | c    |
+-------+------+------+
| china |    1 | NULL |
+-------+------+------+
1 row in set (0.00 sec)
```

### 3. What did you see instead (Required)
```
mysql> insert into t(name, b) values("china", 1), ("sc", 2);
Query OK, 2 rows affected (0.00 sec)
Records: 2  Duplicates: 0  Warnings: 0

mysql> insert into t(name, b) values("cha", 3);
Query OK, 1 row affected (0.00 sec)

mysql> select * from t use index(primary) where name = "china";
Empty set (0.00 sec)
```
### 4. Affected version (Required)

```
Release Version: v4.0.0-beta.2-872-g03003538d
Edition: Community
Git Commit Hash: 03003538d1d7a1762e02fc62365e4832f9df4dd7
Git Branch: master
UTC Build Time: 2020-08-03 03:22:57
GoVersion: go1.13.1
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #18934 

Problem Summary:

For the prefix clustered indices, the original implementation is defective. For example,

```sql
set tidb_enable_clustered_index=1;
create table t(name varchar(255), b int, c int, primary key(name(2)), index idx(b));
```

TiDB stores the clustered index column `name` __intact__, but there is no check against the prefix duplication. 

```sql
insert into t(name, b) values('aaaaa', 1), ('bbbbb', 2);
insert into t(name, b) values('aaa', 3);

Query OK, 1 row affected (0.00 sec)  -- The expected result is "ERROR 1062 (23000): Duplicate entry 'aa' for key 'PRIMARY'"
```

In addition, TiDB does not distinguish __prefix clustered index__ from __prefix plain index__. This results in 2 problems:
- During the index range construction for point-get SQL like `select * from t use index(primary) where name = 'aaaaa';`, the condition value is truncated(`'aaaaa'` is truncated to `'aa'`). In this case, range `['aa', 'aa']` scans nothing.
  ```sql
  select * from t use index(primary) where name = 'aaaaa';
  Empty set (0.00 sec) -- The expected result is 1 row.
  ```
- The column index covering check is not applicable in the clustered index. This requires a double read when a single read is 
adequate:
  ```sql
  explain select /*+ nth_plan(1) */ count(*) from t use index(idx);
  +------------------------------+----------+-----------+-----------------------+--------------------------------+
  | id                           | estRows  | task      | access object         | operator info                  |
  +------------------------------+----------+-----------+-----------------------+--------------------------------+
  | HashAgg_7                    | 1.00     | root      |                       | funcs:count(1)->Column#4       |
  | └─IndexLookUp_16             | 10000.00 | root      |                       |                                |
  |   ├─IndexFullScan_14(Build)  | 10000.00 | cop[tikv] | table:t, index:idx(b) | keep order:false, stats:pseudo |
  |   └─TableRowIDScan_15(Probe) | 10000.00 | cop[tikv] | table:t               | keep order:false, stats:pseudo |
  +------------------------------+----------+-----------+-----------------------+--------------------------------+
  ```

### What is changed and how it works?

What's Changed:
- Check the record prefix duplication when `AddRecord()`.
- Set the `path.commonHandleLens` and `Datasource.commonHandleLens` to `types.UnspecifiedLength`.
- A simple test.

### Related changes

N/A

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU

### Release note 

- No release note

### CODE REVIEW

File: table/tables/tables.go [hunk 1]
```diff
@@ -749,6 +753,56 @@ func (t *TableCommon) AddRecord(sctx sessionctx.Context, r []types.Datum, opts .
 	return recordID, nil
 }
 
+func (t *TableCommon) checkTruncatedClusteredPrimaryKeyDuplication(sctx sessionctx.Context, pkDts []types.Datum,
+	tableCols []*table.Column, pkIdx *model.IndexInfo) error {
+	truncatedDts := make([]types.Datum, len(pkDts))
+	copy(truncatedDts, pkDts)
```
Comment (lzmhhh123) [inline table/tables/tables.go:759]: Can we add another function to check whether the clustered index needs to truncate data? For each record, if we copy its datums, that's a memory heavy operation.

Comment (lzmhhh123) [reply table/tables/tables.go:759]: Or we move this copy after check truncates in the following function.

Comment (tangenta) [reply table/tables/tables.go]: For the prefix clustered index like `primary key(a(2))`, it takes at least two such memory allocations:
1. Truncate the adding record
2. Truncate the target record in table

What's more, it also need to invoke `tables.IterRecord()` to search the target record in the table.

Is the performance overhead unacceptable?

Comment (coocood) [reply table/tables/tables.go]: duplications can be checked by Get operation.

Comment (tangenta) [reply table/tables/tables.go]: We need to check the key's prefix duplication. Does Get operation provide this feature?

Comment (coocood) [reply table/tables/tables.go]: If the primary key is a prefix, the physical key in the storage is prefix too.

**Comment by coocood** [comment]
LGTM

**Pull request merged**